apps/05Mastermind.js

```diff
@@ -1,47 +1,114 @@
 'use strict';
 
 var assert = require('assert');
-var colors = require('colors/safe');
+// var colors = require('colors/safe');
 var prompt = require('prompt');
 prompt.start();
 
 var board = [];
 var solution = '';
+var guess = '';
 var letters = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'];
 
-function printBoard() {
+//TEST solution:
+// solution = 'aaaa';
+// var guess = 'abdc';
+
+function printBoard(hint, guess) {
     for (var i = 0; i < board.length; i++) {
-        console.log(board[i])
+        console.log(board[i]);
+        //there was a missing semicolon on line 18
     }
+    board.push(hint, guess);
 }
 
 function generateSolution() {
     for (var i = 0; i < 4; i++) {
         var randomIndex = getRandomInt(0, letters.length);
         solution += letters[randomIndex];
+
     }
 }
 
 function getRandomInt(min, max) {
   return Math.floor(Math.random() * (max - min)) + min;
 }
 
-function generateHint() {
-    // your code here
+
+function generateHint(solution, guess) {
+  // var hint = 0;
+
+  var solutionArray = solution.split('');
+  // console.log(solutionArray);
+  var guessArray = guess.split('');
+  // console.log(guessArray);
+  var correctLetterLocations = 0;
+      for (var i = 0; i < solutionArray.length; i++)  {
+        if (solutionArray[i] === guessArray[i]) {
+          correctLetterLocations++;
+          solutionArray[i] = null;
+        }
+      }
+      // console.log(solutionArray);
+  var correctLetters = 0;
+  for (var j = 0; j < solutionArray.length; j++)  {
+    if (guessArray.indexOf(solutionArray[j])) {
+      // console.log(guessArray.indexOf(solutionArray[j]));
+      var targetIndex = guessArray.indexOf(solutionArray[j]);
+      if (targetIndex > -1) {
+        // targetIndex = solutionArray[j];
+        // correctLetters++;
+        solutionArray[j] = null;
+        correctLetters++;
+      }
+    }
+  }
+      // console.log('correct letters:  ' + correctLetters);
+      // console.log('correct locations:  ' + correctLetterLocations);
+      // console.log(correctLetters);
+
+  // var hint = colors.red(correctLetterLocations) + "-" + colors.white(correctLetters);
+  var hint = correctLetterLocations + '-' + correctLetters;
+  // console.log(hint);
+  return hint;
+  // board.push(board[guess]);
 }
 
 function mastermind(guess) {
-    // your code here
+
+  if (guess === solution) {
+    // isTrue = true;
+    return 'You guessed it!';
+  }
+  else {
+    // board[board.length] = guess;
+    board += 1;
+    return generateHint(solution, guess);
```

Comment:
Great work on the app! Keep working on your secondattempt.js. I personally am not sure why the colors don't work either. Did you npm install the colors package by the way?